File: pipeline/kite9-pipeline-common/src/main/kotlin/com/kite9/k9server/command/xml/move/ADLMoveCells.java
package com.kite9.k9server.command.xml.move;

import java.util.Collections;
import java.util.List;

import org.kite9.diagram.dom.css.CSSConstants;
import org.kite9.diagram.dom.elements.ADLDocument;
import org.kite9.diagram.dom.elements.StyledKite9XMLElement;
import org.kite9.diagram.dom.managers.IntegerRangeValue;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;
import org.w3c.dom.css.CSSStyleDeclaration;

import com.kite9.k9server.adl.holder.pipeline.ADLDom;
import com.kite9.k9server.command.CommandException;
import com.kite9.k9server.command.xml.AbstractADLCommand;

/**
 * Makes some space in a grid for new cells to be dropped in.
 * 
 * @author robmoffat
 *
 */
public class ADLMoveCells extends AbstractADLCommand {

	public Integer push, from;
	public Boolean horiz;
	public String fragmentId;
	public List<String> excludedIds = Collections.emptyList();
	
	
	public void doMove(ADLDom adl, int push) {
		checkProperties();
		
		ADLDocument doc = adl.getDocument();
		adl.ensureCssEngine(doc);
		Element container = findFragmentElement(doc, fragmentId);	
		int moved = 0;
		
		NodeList contents = container.getChildNodes();
		for (int i = 0; i < contents.getLength(); i++) {
			if (contents.item(i) instanceof StyledKite9XMLElement) {
				StyledKite9XMLElement el = (StyledKite9XMLElement) contents.item(i);
				
				String id = el.getAttribute("id");
				if (!excludedIds.contains(id)) {
					IntegerRangeValue yr = (IntegerRangeValue) el.getCSSStyleProperty(CSSConstants.GRID_OCCUPIES_Y_PROPERTY);
					IntegerRangeValue xr = (IntegerRangeValue) el.getCSSStyleProperty(CSSConstants.GRID_OCCUPIES_X_PROPERTY);
					CSSStyleDeclaration sd = el.getStyle();
					
					if ((xr != null) && (yr != null)) {
						if (horiz) {
							if (xr.getFrom() >= from) {
								xr = new IntegerRangeValue(xr.getFrom() + push, xr.getTo() + push);
								moved++;
							} else if (xr.getTo() >= from) {
								xr = new IntegerRangeValue(xr.getFrom(), xr.getTo() + push);
								moved++;
							}
							
							sd.setProperty(CSSConstants.GRID_OCCUPIES_X_PROPERTY, xr.getCssText(), "");
							el.setComputedStyleMap("", null);  // clears cache
						} else {
							if (yr.getFrom() >= from) {
								yr = new IntegerRangeValue(yr.getFrom() + push, yr.getTo() + push);
								moved++;
							} else if (yr.getTo() >= from) {
								yr = new IntegerRangeValue(yr.getFrom(), yr.getTo() + push);
								moved++;
							}
							
							sd.setProperty(CSSConstants.GRID_OCCUPIES_Y_PROPERTY, yr.getCssText(), "");
							el.setComputedStyleMap("", null);  // clears cache
						}
					}
				}
			}
		}
		LOG.info("Processed move from "+from+" push "+push+" horiz="+horiz+",moved="+moved);
	}

	protected void checkProperties() {
		ensureNotNull("fragmentId", fragmentId);
		ensureNotNull("from", from);
		ensureNotNull("push", push);
	}

	@Override
	public Mismatch applyCommand(ADLDom in) throws CommandException {
		doMove(in, push);
		return null;
	}

	@Override
	public Mismatch undoCommand(ADLDom in) throws CommandException {
		doMove(in, -push);
		return null;
	}
	
}

File: pipeline/kite9-pipeline-common/src/main/kotlin/com/kite9/pipeline/command/xml/move/ADLMoveCells.kt
package com.kite9.pipeline.command.xml.move

import com.kite9.pipeline.adl.holder.pipeline.ADLDom
import com.kite9.pipeline.command.Command
import com.kite9.pipeline.command.CommandContext
import com.kite9.pipeline.command.xml.AbstractADLCommand
import org.kite9.diagram.common.range.BasicIntegerRange
import org.kite9.diagram.common.range.IntegerRange
import org.kite9.diagram.dom.css.CSSConstants
import org.w3c.dom.Document
import org.w3c.dom.Element

/**
 * Makes some space in a grid for new cells to be dropped in.  This is a special
 * operation that actually understands something of the diagram it is being used in.
 *
 * We could change this later to be more like InsertUrlWithChanges...
 *
 * @author robmoffat
 */
class ADLMoveCells : AbstractADLCommand() {

    @JvmField
    var push: Int = 0

    @JvmField
    var from: Int = 0

    @JvmField
    var horiz: Boolean = false

    @JvmField
    var fragmentId: String? = null

    @JvmField
    var excludedIds = emptyList<String>()

    private fun doMove(adl: ADLDom, push: Int, ctx: CommandContext) : Command.Mismatch? {
        checkProperties()
        val doc: Document = adl.document
        val container: Element? = findFragmentElement(doc, fragmentId, ctx)

        if (container == null) {
            return Command.Mismatch { "Couldn't find container $fragmentId" }
        }

        var moved = 0
        val contents = container.childNodes
        for (i in 0 until contents.length) {
            if (contents.item(i) is Element) {
                val el = contents.item(i) as Element
                val id: String = el.getAttribute("id")
                if (!excludedIds.contains(id)) {
                    val r = getRange(el, horiz, ctx)

                    if (r != null) {
                        if (r.from >= from) {
                            setRange(el, horiz, BasicIntegerRange(r.from + push, r.to + push), ctx)
                            moved++
                        } else if (r.to >= from) {
                            setRange(el, horiz, BasicIntegerRange(r.from, r.to + push), ctx)
                            moved++
                        }
                    }
                }
            }
        }

        ctx.log("Processed move from $from push $push horiz=$horiz,moved=$moved")
        return null
    }

    private fun checkProperties() {
        ensureNotNull("fragmentId", fragmentId)
        ensureNotNull("from", from)
        ensureNotNull("push", push)
    }

    override fun applyCommand(d: ADLDom, ctx: CommandContext): Command.Mismatch? {
        return doMove(d, push, ctx)
    }

    override fun undoCommand(d: ADLDom, ctx: CommandContext): Command.Mismatch? {
        return doMove(d, -push, ctx)
    }

    private fun getRange(el: Element, horiz: Boolean, ctx: CommandContext) : IntegerRange? {
        return when (horiz) {
            true -> ctx.getStyleRangeValue(el, CSSConstants.GRID_OCCUPIES_X_PROPERTY)
            false -> ctx.getStyleRangeValue(el, CSSConstants.GRID_OCCUPIES_Y_PROPERTY)
        }
    }

    private fun setRange(el: Element, horiz: Boolean, r: IntegerRange, ctx: CommandContext) {
        when (horiz) {
            true -> ctx.setStyleValue(el, CSSConstants.GRID_OCCUPIES_X_PROPERTY, r.toString())
            false -> ctx.setStyleValue(el, CSSConstants.GRID_OCCUPIES_Y_PROPERTY, r.toString())
        }
    }
}